clang/lib/Sema/SemaOverload.cpp

```diff
@@ -10615,6 +10615,40 @@ bool clang::isBetterOverloadCandidate(
     auto *Guide1 = dyn_cast_or_null<CXXDeductionGuideDecl>(Cand1.Function);
     auto *Guide2 = dyn_cast_or_null<CXXDeductionGuideDecl>(Cand2.Function);
     if (Guide1 && Guide2) {
+      //  -- F1 and F2 are generated from class template argument deduction
+      //  for a class D, and F2 is generated from inheriting constructors
+      //  from a base class of D while F1 is not, ...
+      bool G1Inherited = Guide1->getSourceDeductionGuide() &&
+                         Guide1->getSourceDeductionGuideKind() ==
+                             CXXDeductionGuideDecl::SourceDeductionGuideKind::
+                                 InheritedConstructor;
+      bool G2Inherited = Guide2->getSourceDeductionGuide() &&
+                         Guide2->getSourceDeductionGuideKind() ==
+                             CXXDeductionGuideDecl::SourceDeductionGuideKind::
+                                 InheritedConstructor;
+      if (Guide1->isImplicit() && Guide2->isImplicit() &&
+          G1Inherited != G2Inherited) {
+        const FunctionProtoType *FPT1 =
+            Guide1->getType()->getAs<FunctionProtoType>();
+        const FunctionProtoType *FPT2 =
+            Guide2->getType()->getAs<FunctionProtoType>();
+        assert(FPT1 && FPT2);
+
+        // ... and for each explicit function argument, the parameters of F1 and
+        // F2 are either both ellipses or have the same type
+        if (FPT1->isVariadic() == FPT2->isVariadic() &&
+            FPT1->getNumParams() == FPT2->getNumParams()) {
```

Comment:
This seems wrong compared to the comment -- the comment says we should be looking at each explicit function argument, but this predicate only handles the case where both functions are variadic and have the same number of params.

Am I missing something?

Reply:
Is there a reason we can't just check the types of the functions here instead?  I would assume their prototypes to compare equal.  That said, the wording of `parameters are both ellipses or have the same type` seems intentionally designed for the purpose of excluding the return type (or, perhaps reference/cv specifiers?)?  

If so, I think I'd rather a variant of `hasSameType` that for something like, `hasSameParameterList`.  At that point, we could have IT check for similar variadics as well. We probably want to make sure that `(int, ...)` works too.

Also, I don't think the initial check is right?  Shouldn't that be an 'or' there instead of '&&'?  If both are `NumParams == 0 && isVariadic`, then we want to hit this condition as well.

Reply:
The return types will not be equal as the inherited deduction guides will have some `typename CC<R>::type` as the return type, while non-inherited deduction guides will just have a template specialization type for the class template.

If both functions are variadic and have `getNumParams() == 0`, then that boolean expression evaluates to true right? Can you elaborate on what you think the behavior there should be? I may be misunderstanding what the standards wording on this part is asking.

Reply:
We could assert that `getNumNonObjectParams` is `false` (I don't think we can get explicit objects anywhere, right? Otherwise, the code looks correct, but I agree that something like a "haveSameNonObjectParameters" function would improve readability